include/cetl/span.h

```diff
@@ -1 +1,772 @@
+              typename std::enable_if<std::is_convertible<DeducedElementType (*)[], element_type (*)[]>::value,
+                                      bool>::type = true>
+    constexpr span(const span<DeducedElementType, DeducedExtent>& source) noexcept
+        : data_(source.data())
+        , size_(source.size())
+    {
+    }
+
+    /// Default copy constructor
+    /// @see std::span::span()
+    constexpr span(const span&) noexcept = default;
+
+    ///@}
+    // +----------------------------------------------------------------------+
+    ///@{ @name Operators
+    // +----------------------------------------------------------------------+
+    /// @copydoc span::operator=()
+    ///
+    constexpr span& operator=(const span& rhs) noexcept = default;
+
+    /// @copydoc span::operator[]()
+    ///
+    constexpr reference operator[](size_type idx) const
+    {
+        CETL_DEBUG_ASSERT((idx < size_), "Indexing outside of the span is undefined.");
+        return data_[idx];
+    }
+
+    ///@}
+    // +----------------------------------------------------------------------+
+    ///@{ @name Iterators
+    // +----------------------------------------------------------------------+
+    /// @copydoc span::begin()
+    ///
+    constexpr iterator begin() const noexcept
+    {
+        return iterator(&data_[0]);
+    }
+    /// @copydoc span::end()
+    ///
+    constexpr iterator end() const noexcept
+    {
+        return iterator(&data_[size_]);
+    }
+    /// @copydoc span::rbegin()
+    ///
+    constexpr reverse_iterator rbegin() const noexcept
+    {
+        return reverse_iterator(end());
+    }
+    /// @copydoc span::rend()
+    ///
+    constexpr reverse_iterator rend() const noexcept
+    {
+        return reverse_iterator(begin());
+    }
+
+    // +----------------------------------------------------------------------+
+    ///@{ @name Element Access
+    // +----------------------------------------------------------------------+
+    /// @copydoc span::front()
+    ///
+    constexpr reference front() const
+    {
+        CETL_DEBUG_ASSERT((size_ > 0), "Calling front on an empty span is undefined.");
+        return *begin();
+    }
+
+    /// @copydoc span::back()
+    ///
+    constexpr reference back() const
+    {
+        CETL_DEBUG_ASSERT((size_ > 0), "Calling back on an empty span is undefined.");
+        return *(end() - 1);
+    }
+
+    /// @copydoc span::data()
+    ///
+    constexpr pointer data() const noexcept
+    {
+        return data_;
+    }
+
+    ///@}
+    // +----------------------------------------------------------------------+
+    ///@{ @name Observers
+    // +----------------------------------------------------------------------+
+
+    ///
+    /// Returns if the span has a zero size or not.
+    /// @return true if the span size is 0.
+    [[nodiscard]] constexpr bool empty() const noexcept
+    {
+        return (size_ == 0);
+    }
+
+    ///
+    /// Returns the size of the span.
```

Comment:
it irks me that they didn't call this count() in the standard.

Reply:
sorry

Reply:
Not your fault. I probably need to write my own standard library, I'll call it the `grumpy template library` or `gtl::`

Reply:
I'll be your first customer.